### PR TITLE
Add batch actions to all/managed works page.

### PR DESCRIPTION
**Pull request by jonathandixon**
Adds ability to batch select works on the All/Managed page (dashboard/works) and perform batch operations: delete, add to a collection, and edit.

@samvera/hyrax-code-reviewers

### CODE REVIEW

**Comment by elrayle** [comment]
This addresses issue #2280 - Make read access works available to add to collections

**Comment by jonathandixon** [comment]
Added suggested tests: batch edit/delete that includes read-only works updates/deletes the edit access works and not the read access works.

File: spec/views/hyrax/my/_search_header.html.erb_spec.rb [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
   context "on my works page" do
     before do
+      view.lookup_context.prefixes.push "hyrax/my/works"
       allow(view).to receive(:on_my_works?).and_return(true)
```
Comment (jcoyne) [inline spec/views/hyrax/my/_search_header.html.erb_spec.rb]: Remove

File: spec/views/hyrax/my/_search_header.html.erb_spec.rb [hunk 2]
```diff
@@ -22,15 +23,14 @@
     end
   end
 
-  context "not on my works page (i.e. one of the collections index pages)" do
+  context "on my collections page" do
     before do
+      view.lookup_context.prefixes.push "hyrax/my/collections"
       allow(view).to receive(:on_my_works?).and_return(false)
```
Comment (jcoyne) [inline spec/views/hyrax/my/_search_header.html.erb_spec.rb]: remove?

**Comment by jonathandixon** [comment]
@jcoyne I'm fine with removing `DashboardHelperBehavior#on_my_works?`. Should this be done in a separate issue? I worry that removing it here will get lost and that downstream applications that use it in their templates will have problems on upgrade.

**Comment by jcoyne** [comment]
@jonathandixon I think you should just remove it here because we have no other uses of it in hyrax.  If you don't want to  remove it here out of an abundance of caution, you should at least deprecate it.

File: app/views/hyrax/my/works/_tabs.html.erb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <ul class="nav nav-tabs" id="my_nav" role="navigation">
   <span class="sr-only">You are currently listing your works.  You have <%= @response.docs.count %> <%= 'work'.pluralize(@response.docs.count)%> </span>
   <li<%= ' class="active"'.html_safe if current_page?(hyrax.dashboard_works_path(locale: nil)) %>>
-    <%= link_to t("hyrax.dashboard.#{current_ability.admin? ? 'all' : 'managed'}.works"), hyrax.dashboard_works_path %>
+    <%= link_to t("hyrax.dashboard.#{current_ability.admin? ? 'all' : 'managed'}.works"), hyrax.dashboard_works_path, data: { turbolinks: false } %>
```
Comment (jcoyne) [inline app/views/hyrax/my/works/_tabs.html.erb]: Why are you doing `turbolinks: false` here?

File: app/views/hyrax/my/works/_batch_actions.html.erb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<div class="batch-info">
+  <%= render 'hyrax/dashboard/collections/form_for_select_collection', user_collections: @user_collections %>
+
+  <div class="batch-toggle">
+    <% session[:batch_edit_state] = "on" %>
+    <div class="button_to-inline">
+      <%= button_to "Edit Selected", hyrax.edit_batch_edits_path, method: :get, class: "btn btn-update btn-primary hidden submits-batches", data: { behavior: 'batch-edit' }, id: 'batch-edit' %>
```
Comment (jcoyne) [inline app/views/hyrax/my/works/_batch_actions.html.erb]: Can you add this string to i18n? and on line 10?

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 5]
```diff
@@ -35,20 +37,27 @@
...
       expect(controller).to receive(:can?).with(:edit, two.id).and_return(true)
+      expect(controller).to receive(:can?).with(:edit, three.id).and_return(false)
     end
 
     it "is successful" do
       put :update, params: { update_type: "delete_all" }
       expect(response).to redirect_to(dashboard_path(locale: 'en'))
       expect { GenericWork.find(one.id) }.to raise_error(Ldp::Gone)
       expect { GenericWork.find(two.id) }.to raise_error(Ldp::Gone)
+      expect(GenericWork.find(three.id).id).to eq(three.id)
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: This should be `expect(GenericWork).to exist(three.id)` (or alternately `expect(GenericWork.exists?(three.id).to be true`)

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 6]
```diff
@@ -61,13 +70,15 @@
       expect(response).to be_redirect
       expect(GenericWork.find(one.id).subject).to eq ["zzz"]
       expect(GenericWork.find(two.id).subject).to eq ["zzz"]
+      expect(GenericWork.find(three.id).subject).not_to eq ["zzz"]
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: Negative assertions are not nearly as specific as positive assertions.  Is there a reason you can't do `expect(GenericWork.find(three.id).subject).to eq ["whavever value it initially had"]`?

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 7]
```diff
@@ -61,13 +70,15 @@
...
     end
 
     it "updates permissions" do
       put :update, params: { update_type: "update", visibility: "authenticated" }
       expect(response).to be_redirect
       expect(GenericWork.find(one.id).visibility).to eq "authenticated"
       expect(GenericWork.find(two.id).visibility).to eq "authenticated"
+      expect(GenericWork.find(three.id).visibility).not_to eq "authenticated"
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: Can we assert what we expect it to be? `restricted`?

File: spec/controllers/hyrax/dashboard/collections_controller_spec.rb [hunk 8]
```diff
@@ -74,7 +74,7 @@
 
       it "adds docs to the collection and adds the collection id to the documents in the collection" do
         post :create, params: {
-          batch_document_ids: [asset1.id],
+          batch_document_ids: [asset1.id, unowned_asset.id],
```
Comment (jcoyne) [inline spec/controllers/hyrax/dashboard/collections_controller_spec.rb]: Why are you doing this change?  It seems odd that no assertions are changing.

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 9]
```diff
@@ -35,20 +37,27 @@
     let!(:two) do
       create(:generic_work, creator: ["Fred"], title: ["abc"], language: ['en'], user: user)
     end
+
+    let!(:three) do
+      create(:generic_work, creator: ["Fred"], title: ["abc"], language: ['en'])
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: On line 13 above you are using `create(:work,...)`  can we stick with that for consistency?

File: app/views/hyrax/dashboard/works/_list_works.html.erb [hunk 10]
```diff
@@ -1,4 +1,8 @@
 <tr id="document_<%= document.id %>">
+  <td>
+    <label for="batch_document_<%= document.id %>" class="sr-only"><%=t("hyrax.dashboard.my.sr.batch_checkbox")%></label>
+    <%= render 'hyrax/batch_select/add_button', document: document %>&nbsp;
```
Comment (jcoyne) [inline app/views/hyrax/dashboard/works/_list_works.html.erb]: What's the point of the `&nbsp` here?

File: app/views/hyrax/my/works/_tabs.html.erb (hunk 3 above)
Comment (jonathandixon) [reply app/views/hyrax/my/works/_tabs.html.erb]: This relates to issue #1191. Without it, you end up with an `ActionController::InvalidAuthenticityToken` error.

File: spec/controllers/hyrax/dashboard/collections_controller_spec.rb (hunk 8 above)
Comment (jonathandixon) [reply spec/controllers/hyrax/dashboard/collections_controller_spec.rb]: This is to verify that only works the user has access to are added to the collection. This passes an ID of an unowned work with the request but that shouldn't change the original expectation that only `asset1` is added to the collection.

File: app/views/hyrax/dashboard/works/_list_works.html.erb (hunk 10 above)
Comment (jonathandixon) [reply app/views/hyrax/dashboard/works/_list_works.html.erb]: This was a copy and paste. It appears the `&nbsp;` is used to manipulate the vertical position of the checkbox inside the table cell (without it the checkbox renders below verticle center).

**Pull request merged**